examples/throughput/publisher.cpp

```diff
@@ -0,0 +1,223 @@
+  done = true;
+}
+
+template<typename T>
+bool wait_for_reader(dds::pub::DataWriter<T> &writer)
+{
+  std::cout << "\n" << pubprefix << "Waiting for a reader...\n" << std::flush;
+
+  dds::core::cond::StatusCondition sc = dds::core::cond::StatusCondition(writer);
+  sc.enabled_statuses(dds::core::status::StatusMask::publication_matched());
+
+  dds::core::cond::WaitSet waitset;
+  waitset.attach_condition(sc);
+
+  dds::core::cond::WaitSet::ConditionSeq conditions =
+    waitset.wait(dds::core::Duration::from_secs(30));
+
+  for (const auto & c:conditions)
+  {
+    if (c == sc)
+      return true;
+  }
+  
+  std::cout << pubprefix << "Did not discover a reader.\n" << std::flush;
+
+  return false;
+}
+
+template<typename T>
+void start_writing(dds::pub::DataWriter<T> &writer, T &sample,
+    std::chrono::milliseconds &d_int, uint32_t b_size, std::chrono::seconds &t_out)
+{
+  bool timedOut = false;
+
+  dds::pub::AnyDataWriter wr(writer);
+  /*wr.set_batch(true);  currently the C++ binding does not support batched writing*/
+
+  auto pubStart = std::chrono::steady_clock::now();
+  auto reportstart = pubStart;
+
+  if (!done)
+  {
+    std::cout << pubprefix << "Writing samples" << std::flush;
+
+    while (!done && !timedOut)
+    {
+      auto burstStart = std::chrono::steady_clock::now();
+      for (uint32_t i = 0; i < b_size; i++)
+      {
+        try {
+          writer.write(sample);
+        } catch (const dds::core::TimeoutError &) {
+          timedOut = true;
+        } catch (const dds::core::Exception &e) {
+          std::cout << "\n" << pubprefix "Writer encountered the following error: \"" << e.what() << "\".\n" << std::flush;
+          done = true;
+        } catch (...) {
+          std::cout << "\n" << pubprefix "Writer encountered an error.\n" << std::flush;
+          done = true;
+        }
+        sample.count()++;
+      }
+
+      /*wr.write_flush();  currently the C++ binding does not support batched writing*/
+      std::this_thread::sleep_until(burstStart + d_int);
+
+      auto n = std::chrono::steady_clock::now();
+      if (t_out > std::chrono::milliseconds(0) &&
+          n > pubStart+t_out)
+      {
+        timedOut = true;
+      } else if (reportstart + std::chrono::seconds(1) < n) {
+        std::cout << "." << std::flush;
+        reportstart = n;
+      }
+
+    if (writer.publication_matched_status().current_count() == 0)
+      timedOut = true;
+    }
+
+    std::cout << "\n" << pubprefix << (done ? "Terminated" : "Timed out") << ", " << sample.count() << " samples written.\n" << std::flush;
+  }
+  /*wr.write_flush();  currently the C++ binding does not support batched writing*/
+}
+
+int main (int argc, char **argv)
+{
+  ThroughputModule::DataType sample;
+
+  if (parse_args(argc, argv) == EXIT_FAILURE) {
+    return EXIT_FAILURE;
+  }
+
+  dds::domain::DomainParticipant participant(domain::default_id());
+
+  dds::topic::Topic<ThroughputModule::DataType> topic(participant, "Throughput");
```

Comment:
I chose this because this completely mirrors the QoS settings done in the C throughput examples, where the topic has the default QoS set, while the writer has the explicit QoS settings.
These choices for QoS etc. make the C and C++ examples interchangeable (e.g. having a C++ throughput publisher and a C subscriber).

Reply:
Seems like a reason to modify the C example as well in that case. If our examples are to show best-practices, then we shouldn't start by not setting the Qos on the topics ourselves.

Reply:
@e-hndrks I'll take a look at the C examples and update them if needed so they match this style.